### ISSUE TITLE
Failed update to 8.0.2

### ISSUE THREAD

**Issue by dolphinict** [owncloud/core#14916]
Hi Folks,

I have 8.0.2 on my owncloud install, upgrading from 8.0. When I go to the web interface to do the database upgrade it errors saying:

"An exception occurred while executing 'CREATE TABLE oc_ownc_appconfig_rj2l56no8qhw6 (`appid` VARCHAR(32) DEFAULT '' NOT NULL, `configkey` VARCHAR(64) DEFAULT '' NOT NULL, `configvalue` LONGTEXT DEFAULT NULL, INDEX oc_wdpw5tnh1wdyr (`configkey`), INDEX oc_oazv70agn2u7q (`appid`), PRIMARY KEY(`appid`, `configkey`)) DEFAULT CHARACTER SET utf8 COLLATE utf8_bin ENGINE = InnoDB': SQLSTATE[42S01]: Base table or view already exists: 1050 Table 'oc_ownc_appconfig_rj2l56no8qhw6' already exists"

I went into the database and got puzzled. There are no tables beginning oc_ownc at all. There is ownc_appconfig but no prefix or suffix as in the MySQL query output as an error.

I tried dropping back to a backup of 8.0 and got the Start update message which fails with the same error. Net result, I have no owncloud now.

**Comment by nickvergessen** [comment]
Please use the issue template:
https://github.com/owncloud/core/blob/master/CONTRIBUTING.md#tldr

Especially server and database versions as well as the logs are important

**Comment by dolphinict** [comment]
I can see a couple of things. The updater thinks it is upgrading from 7.0.4.2 when I was already on 8.0 which might be causing it to try and recreate things. 

The first error looks to be relating to loading the list of apps, so it is possible that an app is causing this but I can't get into owncloud to do anything about that.

### Steps to reproduce
1. Visit owncloud URL and receive "ownCloud will be updated to version 8.0.2"
2. Click "Start update"

### Expected behaviour
Update should complete and go to login page

### Actual behaviour
Upgrade fails with SQL error output to screen, error in browser log below.

### Server configuration
**Operating system**: CentOS 6.6

**Web server:** Apache/2.2.29

**Database:** MySQL 5.6.23 

**PHP version:** PHP 5.4.34

**ownCloud version:** 8.0.2

**Updated from an older ownCloud or fresh install:**
Updated from 8.0

**List of activated apps:**
These were enabled originally but I have set enabled to no in the db manually. There are no apps with enabled=yes any more
files
gallery
files_sharing
files_trashbin
files_versions
files_pdfviewer
files_texteditor

**The content of config/config.php:**

```
<?php
$CONFIG = array (
  'passwordsalt' => '',
  'datadirectory' => '/home/username/public_html/cloud/data',
  'dbtype' => 'mysql',
  'version' => '7.0.4.2',
  'dbname' => 'username_ownc1',
  'dbhost' => 'localhost',
  'dbtableprefix' => 'ownc_',
  'dbuser' => 'username_ownc1',
  'dbpassword' => '',
  'installed' => true,
  'versions' => 'true',
  'instanceid' => '5077620f6250c',
  'maintenance' => false,
  'forcessl' => true,
  'maxZipInputSize' => 838860800,
  'allowZipDownload' => true,
  'theme' => '',
  'trusted_domains' =>
  array (
    0 => 'www.domainname.co.uk',
  ),
  'secret' => '',
);
```

**Are you using external storage, if yes which one:** no

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** no


```
sqlite> select * from oc_appconfig where appid='user_ldap';
Error: no such table: oc_appconfig

data/owncloud.db is a zero byte file
```

### Client configuration
**Browser:** Tested in Chrome/IE/Firefox

**Operating system:** Win7/8 tested

### Logs
#### Web server error log
```
No apache error log generated at time of failure.
```

#### ownCloud log (data/owncloud.log)
```
{"reqId":"03d2d1c729d909669e5f31d0cd351813","remoteAddr":"1.1.1.1","app":"index","message":"Exception: {\"Message\":\"\",\"Code\":0,\"Trace\":\"#0 \\\/home\\\/username\\\/public_html\\\/cloud\\\/lib\\\/private\\\/app.php(77): OC_App::loadApp('files_sharing')\\n#1 \\\/home\\\/username\\\/public_html\\\/cloud\\\/lib\\\/private\\\/util.php(75): OC_App::loadApps(Array)\\n#2 \\\/home\\\/username\\\/public_html\\\/cloud\\\/lib\\\/base.php(766): OC_Util::setupFS()\\n#3 \\\/home\\\/username\\\/public_html\\\/cloud\\\/index.php(36): OC::handleRequest()\\n#4 {main}\",\"File\":\"\\\/home\\\/username\\\/public_html\\\/cloud\\\/lib\\\/private\\\/app.php\",\"Line\":96}","level":4,"time":"2015-03-16T12:28:15+00:00"}
{"reqId":"c50bfc73fe1abe4ce8be2fe7c60d1729","remoteAddr":"1.1.1.1","app":"core","message":"Simulated database structure update failed (exception 'OC\\DB\\MigrationException' with message 'An exception occurred while executing 'CREATE TABLE oc_ownc_appconfig_bzapajnqc96gy (`appid` VARCHAR(32) DEFAULT '' NOT NULL, `configkey` VARCHAR(64) DEFAULT '' NOT NULL, `configvalue` LONGTEXT DEFAULT NULL, INDEX oc_27rq253b9pfvk (`configkey`), INDEX oc_qagjht18snkm7 (`appid`), PRIMARY KEY(`appid`, `configkey`)) DEFAULT CHARACTER SET utf8 COLLATE utf8_bin ENGINE = InnoDB':\n\nSQLSTATE[42S01]: Base table or view already exists: 1050 Table 'oc_ownc_appconfig_bzapajnqc96gy' already exists' in \/home\/username\/public_html\/cloud\/lib\/private\/db\/migrator.php:133\nStack trace:\n#0 \/home\/username\/public_html\/cloud\/lib\/private\/db\/migrator.php(91): OC\\DB\\Migrator->checkTableMigrate(Object(Doctrine\\DBAL\\Schema\\Table))\n#1 \/home\/username\/public_html\/cloud\/lib\/private\/db\/mdb2schemamanager.php(113): OC\\DB\\Migrator->checkMigrate(Object(Doctrine\\DBAL\\Schema\\Schema))\n#2 \/home\/username\/public_html\/cloud\/lib\/private\/db.php(252): OC\\DB\\MDB2SchemaManager->simulateUpdateDbFromStructure('\/home\/username\/p...')\n#3 \/home\/username\/public_html\/cloud\/lib\/private\/updater.php(238): OC_DB::simulateUpdateDbFromStructure('\/home\/username\/p...')\n#4 \/home\/username\/public_html\/cloud\/lib\/private\/updater.php(211): OC\\Updater->checkCoreUpgrade()\n#5 \/home\/username\/public_html\/cloud\/lib\/private\/updater.php(150): OC\\Updater->doUpgrade('8.0.2.0', '7.0.4.2')\n#6 \/home\/username\/public_html\/cloud\/core\/ajax\/update.php(50): OC\\Updater->upgrade()\n#7 {main})","level":4,"time":"2015-03-16T12:28:17+00:00"}
```

#### Browser log
```
Owncloud outputs error to screen as follows:
Updating ownCloud to version 8.0.2, this may take a while.

Turned on maintenance mode
An exception occurred while executing 'CREATE TABLE oc_ownc_appconfig_ny7i4kvghy513 (`appid` VARCHAR(32) DEFAULT '' NOT NULL, `configkey` VARCHAR(64) DEFAULT '' NOT NULL, `configvalue` LONGTEXT DEFAULT NULL, INDEX oc_rikci0ac2jjpq (`configkey`), INDEX oc_ybw5bdvcnzdsh (`appid`), PRIMARY KEY(`appid`, `configkey`)) DEFAULT CHARACTER SET utf8 COLLATE utf8_bin ENGINE = InnoDB': SQLSTATE[42S01]: Base table or view already exists: 1050 Table 'oc_ownc_appconfig_ny7i4kvghy513' already exists
The update was unsuccessful. Please report this issue to the ownCloud community.
```

**Comment by DeepDiver1975** [comment]
This issue has been reported multiple times already - time for us to finally have a look.

**Comment by dolphinict** [comment]
Let me know if I can pull any more info for you or run any tests to help out.

**Comment by karlitschek** [comment]
@DeepDiver1975 agreed. who can help with that?

**Comment by icewind1991** [comment]
The problem seems to come down to dbal generating the wrong diff because it somehow doesn't find our newly created table when listing the existing tables.

No idea how to further debug that or fix it without being able to reproduce it locally

**Comment by dolphinict** [comment]
I can make modifications to the code for testing or do anything necessary to assist. I'd not a full PHP dev but have used it a fair bit for projects from time to time. Just let me know what you need.

**Comment by jeekajoo** [comment]
same here..
```
An exception occurred while executing 'CREATE TABLE oc_agreattablesuffix_appconfig_ekiuenjs1muye (`appid` VARCHAR(32) DEFAULT '' NOT NULL, `configkey` VARCHAR(64) DEFAULT '' NOT NULL, `configvalue` LONGTEXT DEFAULT NULL, INDEX oc_5l310lpqpdp76 (`configkey`), INDEX oc_o3dc73of6ppb8 (`appid`), PRIMARY KEY(`appid`, `configkey`)) DEFAULT CHARACTER SET utf8 COLLATE utf8_bin ENGINE = InnoDB': SQLSTATE[42S01]: Base table or view already exists: 1050 Table 'oc_agreattablesuffix_appconfig_ekiuenjs1muye' already exists
```
using 8.0.2-5 from https://software.opensuse.org/download/package?project=isv:ownCloud:community&package=owncloud
let me know how I can help

**Comment by oparoz** [comment]
And this happens every time there is an app update as well.

**Comment by jeekajoo** [comment]
@oparoz thanks man, forgot this one.

**Comment by DeepDiver1975** [comment]
> The problem seems to come down to dbal generating the wrong diff because it somehow doesn't find our newly created table when listing the existing tables.

do we cache these information somehow?

**Comment by dolphinict** [comment]
--skip-migration-test does not resolve it for me unfortunately. I get the following output which makes it look like the script is detecting that it isn't being called from the web browser

Status: 302 Moved Temporarily
X-Powered-By: PHP/5.4.34
Set-Cookie: 5077620f6250c=dede105d40a23b581c08a2af198e0da8; path=/; HttpOnly
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Pragma: no-cache
Strict-Transport-Security: max-age=31536000
Location: https://www.mydomainname.tld
Content-type: text/html

**Comment by dolphinict** [comment]
I've figured out why I couldn't run occ myself. The server was limiting the user access so sudo as the user was working but the script wasn't. I removed those restrictions and was able to run php occ upgrade --skip-migration-test so I can't help with the original problem any more.

Sorry but I needed OC back up and running!

**Comment by SierraBMWCCA** [comment]
I'm a newbie and having the same problem. I first tried to upgrade through Softaculous but it would not let me upgrade from 7.0.4 to 8.0.2 so I went to the ownCloud app and tried to upgrade but received the same error (couldn't find this error so I created one: Error upgrading from 7.0.4 to 8.0.2 #15872). I have tried updating several times, disabling 3rd Party apps - I disabled most of the apps - and still receive the same error. I am at a loss at what to do next. Being a newbie do you have instructions on how to manually upgrade? I'm not sure where I would go in my hosting control panel to run some of the instructions I was reading. Any help would be appreciated.

**Comment by TraceyC77** [comment]
I've also just encountered this error, and wanted to provide additional information that might help in troubleshooting & resolving the issue. I'm happy to provide additional information and testing that might be needed.

### Steps to reproduce
1. Installatron updated OwnCloud on the server from 8.0.2 to 8.0.3
2. Visited the URL for the OwnCloud interface
3. Clicked the Start Update button

### Expected behaviour
OwnCloud is updated

### Actual behaviour
The update failed with an exception, OwnCloud cannot be accessed via the web interface.

### Server configuration
**Operating system**: CentOS 6.6

**Web server:** Apache/2.4.10 (Unix)

**Database:** MySQL 5.5.42

**PHP version:** 5.5.18

**ownCloud version:** 8.0.2

**Updated from an older ownCloud or fresh install:** Upgraded

**List of activated apps:** Gallery, Files, Calendar, Contacts

**The content of config/config.php:**

```
<?php                                                                                       [11.48.4 (build 2)]
$CONFIG = array (
  'instanceid' => 'ocae55aec19d',
  'passwordsalt' => 'x',
  'trusted_domains' => 
  array (
    0 => 'www.tlcnet.info',
  ),
  'datadirectory' => '/home/tlcnet/public_html/owncloud/data',
  'dbtype' => 'mysql',
  'version' => '8.0.2.0',
  'dbname' => 'tlcnet_ownc1',
  'dbhost' => 'localhost',
  'dbtableprefix' => 'ownc_',
  'dbuser' => 'x',
  'dbpassword' => 'x',
  'installed' => true,
  'mail_smtpmode' => 'php',
  'mail_from_address' => 'owncloud',
  'mail_domain' => 'tlcnet.info',
  'mail_smtpauthtype' => 'PLAIN',
  'mail_smtpauth' => true,
  'mail_smtphost' => 'mail.tlcnet.info',
  'forcessl' => true,
  'maintenance' => false,
  'secret' => 'x',
  'forceSSLforSubdomains' => true,
  'theme' => '',                                                                                                     
);    
```

**Are you using external storage, if yes which one:** no

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** Webdav

### Client configuration
**Browser:** Chromium Version 41.0.2272.76

**Operating system:** Linux Mint 17.1

### Logs
#### Web server error log
```
No apache error log generated at time of failure.
```

#### ownCloud log (data/owncloud.log)
```
{"reqId":"5a3fc80b0b75726b0a800be1703fb6d8","remoteAddr":"208.74.121.102","app":"core","message":"Simulated database structure update failed (exception 'OC\\DB\\MigrationException' with message 'An exception occurred while executing 'CREATE TABLE oc_ownc_appconfig_mq1vdy4i38b2g (`appid` VARCHAR(32) DEFAULT '' NOT NULL, `configkey` VARCHAR(64) DEFAULT '' NOT NULL, `configvalue` LONGTEXT DEFAULT NULL, INDEX oc_hewohuzh8t9m9 (`configkey`), INDEX oc_vu1qsf9ph51rw (`appid`), PRIMARY KEY(`appid`, `configkey`)) DEFAULT CHARACTER SET utf8 COLLATE utf8_bin ENGINE = InnoDB':\n\nSQLSTATE[42S01]: Base table or view already exists: 1050 Table 'oc_ownc_appconfig_mq1vdy4i38b2g' already exists' in \/home\/tlcnet\/public_html\/owncloud\/lib\/private\/db\/migrator.php:133\nStack trace:\n#0 \/home\/tlcnet\/public_html\/owncloud\/lib\/private\/db\/migrator.php(91): OC\\DB\\Migrator->checkTableMigrate(Object(Doctrine\\DBAL\\Schema\\Table))\n#1 \/home\/tlcnet\/public_html\/owncloud\/lib\/private\/db\/mdb2schemamanager.php(113): OC\\DB\\Migrator->checkMigrate(Object(Doctrine\\DBAL\\Schema\\Schema))\n#2 \/home\/tlcnet\/public_html\/owncloud\/lib\/private\/db.php(239): OC\\DB\\MDB2SchemaManager->simulateUpdateDbFromStructure('\/home\/tlcnet\/pu...')\n#3 \/home\/tlcnet\/public_html\/owncloud\/lib\/private\/updater.php(254): OC_DB::simulateUpdateDbFromStructure('\/home\/tlcnet\/pu...')\n#4 \/home\/tlcnet\/public_html\/owncloud\/lib\/private\/updater.php(226): OC\\Updater->checkCoreUpgrade()\n#5 \/home\/tlcnet\/public_html\/owncloud\/lib\/private\/updater.php(150): OC\\Updater->doUpgrade('8.0.3.4', '8.0.2.0')\n#6 \/home\/tlcnet\/public_html\/owncloud\/core\/ajax\/update.php(58): OC\\Updater->upgrade()\n#7 {main})","level":4,"time":"2015-05-11T13:34:17+00:00"}

```

#### Browser log
```
Error that is output to screen after clicking Start:

An exception occurred while executing 'CREATE TABLE oc_ownc_appconfig_nsto2qcghxpaf (`appid` VARCHAR(32) DEFAULT '' NOT NULL, `configkey` VARCHAR(64) DEFAULT '' NOT NULL, `configvalue` LONGTEXT DEFAULT NULL, INDEX oc_e8ivtrug3fhm5 (`configkey`), INDEX oc_2f2ovonlc41bk (`appid`), PRIMARY KEY(`appid`, `configkey`)) DEFAULT CHARACTER SET utf8 COLLATE utf8_bin ENGINE = InnoDB': SQLSTATE[42S01]: Base table or view already exists: 1050 Table 'oc_ownc_appconfig_nsto2qcghxpaf' already exists
```

**Comment by oparoz** [comment]
This seems to be fixed for me on 8.1 (PHP5.5, MariaDB, Redis) while it still fails on 8.0.3, so maybe the milestone should be adjusted? @DeepDiver1975

**Comment by TraceyC77** [comment]
I have, as a workaround, updated my installation on command line with 
`# ./occ upgrade --skip-migration-test`

Sorry for being unable to leave the server in the broken state, but I'm not able to have a long outage to the calendar and card services.

**Comment by PVince81** [comment]
Are you guys by any chance all using a custom prefix ? I see "ownc" in @dolphinict's config.php but the table name is still "oc_ownc". It looks like something might be wrong.

I'll try it locally with that prefix.

**Comment by oparoz** [comment]
But the error message says "Base table or view already exists". So it seems there is a conflict, not a missing table, so I don't think the prefix plays a role here.

**Comment by PVince81** [comment]
Even then, I need a starting point by finding what the configs have in common.
Both @TraceyC77 and @dolphinict have a non-standard prefix. Who knows what can go wrong...

**Comment by oparoz** [comment]
I use custom prefixes, but I didn't have a problem on 8.1, so unless I got lucky, a commit must have fixed it already.

**Comment by PVince81** [comment]
:bomb: problem reproduced:
```
An exception occurred while executing 'CREATE TABLE oc_ownc_appconfig_6y8djb4lztkis (`appid` VARCHAR(32) DEFAULT '' NOT NULL, `configkey` VARCHAR(64) DEFAULT '' NOT NULL, `configvalue` LONGTEXT DEFAULT NULL, INDEX oc_8qw0sdngasvgj (`configkey`), INDEX oc_dur9b7njllrhr (`appid`), PRIMARY KEY(`appid`, `configkey`)) DEFAULT CHARACTER SET utf8 COLLATE utf8_bin ENGINE = InnoDB': SQLSTATE[42S01]: Base table or view already exists: 1050 Table 'oc_ownc_appconfig_6y8djb4lztkis' already exists
```

**Comment by PVince81** [comment]
Tried again without prefix and upgrading works.

I can confirm that it's related to the prefix.
Next up: debugging session

**Comment by oparoz** [comment]
I can see that the INDEX is missing the proper prefix.

**Comment by nickvergessen** [comment]
@PVince81 https://github.com/owncloud/core/pull/14658 done

**Comment by nickvergessen** [comment]
As per https://github.com/owncloud/core/pull/14658 indexes dont contain the prefix atm and also the current indexes dont contain it. So of course it will fail to create the same table with a different name, because the index name is non-unique

**Comment by PVince81** [comment]
Can be reproduced this way:
1) Edit autotest.sh and change the mysql prefix to "ownc_"
2) Run `./autotest.sh mysql tests/lib/db/migrator.php`

**Comment by PVince81** [comment]
But even it they don't, why does the upgrade fail ? This is not a situation where we'd have conflicts of prefixes.

**Comment by PVince81** [comment]
I just tried https://github.com/owncloud/core/pull/14658 and it does not fix the issue.

**Comment by nickvergessen** [comment]
yes, because `*prefix*` is missing on all index names in our schema.xml

**Comment by PVince81** [comment]
Okay, so does it mean that when simulating the DB upgrade, it would try and recreate indices with the exact same names ? If yes, then why does it work when the prefix is "oc_" ?

**Comment by nickvergessen** [comment]
There are also reports with oc_ https://github.com/owncloud/core/issues/15416 ?
Also I just found https://github.com/owncloud/core/blob/master/lib/private/db/migrator.php#L167-L173

**Comment by PVince81** [comment]
In my case the problem happens when I use "ownc_" but works properly when using "oc_".
Could be yet another issue :frowning:

**Comment by PVince81** [comment]
I'll first debug the case I have in hand, then we can see if we can find other clues.

**Comment by PVince81** [comment]
It looks like the diff code doesn't detect that the table already exists, so it tries to recreate it.
I'm comparing the code paths for the "oc_" and "ownc_" approaches now.

**Comment by PVince81** [comment]
With "oc_":
```
    $sourceSchema->_tables['oc_autotest.oc_oc_test_ol2r32oz6lln0_xxhzfb3q76hdg'] = (object|Doctrine\DBAL\Schema\Table[10])+;
    $targetSchema->_tables['oc_autotest.oc_oc_test_ol2r32oz6lln0_xxhzfb3q76hdg'] = (object|Doctrine\DBAL\Schema\Table[10])+;
```

With "ownc_":
```
    $sourceSchema->_tables['oc_autotest.ownc_test_o64zqpk4advpg'] = (object|Doctrine\DBAL\Schema\Table[10])+;
    $targetSchema->_tables['oc_autotest.oc_ownc_test_o64zqpk4advpg_9fapearyv0d7z'] = (object|Doctrine\DBAL\Schema\Table[10])+;
```

If I remember well, Doctrine might have a global attribute for table prefixes, maybe it isn't set properly.

**Comment by PVince81** [comment]
Hmm I wonder if the fact that we hard-coded "oc_" . $name makes Doctrine find the tables properly (the ones called "oc_oc_" vs "oc_ownc_" due to the fact that the hard-coded "oc_" matches the prefix.
So basically being lucky that it works thanks to that.

**Comment by PVince81** [comment]
Yep... replacing the hard-coded "oc_" with the actual table prefix solves it.
I'll prepare a PR.

**Comment by PVince81** [comment]
How nice, @martin-rueegg already fixed this on master: https://github.com/owncloud/core/pull/14660

I'll still provide a PR with my unit test.

**Comment by nickvergessen** [comment]
ah, should propably backport aswell?

**Comment by PVince81** [comment]
Backport requested here: https://github.com/owncloud/core/pull/14660#issuecomment-103087274

**Comment by PVince81** [comment]
I'll prepare the PR, it should be low risk anyway (at least for people using the default prefix)

**Comment by PVince81** [comment]
Backport PR here: https://github.com/owncloud/core/pull/16421

**Comment by MorrisJobke** [comment]
This will be fixed in 8.0.4 and 8.1.0
